site/content/blog/release-0.30.0.md

```diff
@@ -0,0 +1,134 @@
+please get them into us now so your needs can be met.
+
+Without further ado, here is the upgrade guide. Enjoy!
+
+## Re-exports
+
+We are trying to separate the API from the directory structure, as part of this we are attempting to
+only re-export from the crate root types that exist in the standard Bitcoin vernacular i.e., for
+pubkey you can `use bitcoin::PublicKey` but to get an x-only pubkey you need to go to the `key`
+module `use bitcoin::key::XOnlyPublicKey`.
+
+Please note this is still work-in-progress, suggestions welcome.
+
+## Code moves
+
+We moved _a lot_ of stuff around. This was a precursor to crate smashing which we have now started.
+Hopefully things are in intuitive places, it might be useful to take a quick look at the new
+directory structure to get a feel for things, specifically:
+
+- We have a workspace now! The main crate now lives in `bitcoin/`. The `bitcoin_hashes` repository
+  has been merged into the `rust-bitcoin` repository and now lives under `hashes/`. There is a new
+  `hex` crate so you don't need to depend on `hashes` for hex stuff anymore. There is an `internals`
+  crate meant for internal use.
+
+- The `util` module is all but gone, try just removing `util::` at first, most modules are
+  re-exported at the crate root.
+
+- Cryptography related stuff is now in the `crypto` module, all submodules are re-exported so
+  you can import them directly (e.g. `use bitcoin::ecdsa`). Modules include: (`ecdsa`, `taproot`,
+  `sighash`, `key`)
+
+- Some hash wrapper types are now to be found in the module that they are used in e.g.,
+  `TapLeafHash` is in the `taproot` module. Others are still in `hash_types` but the re-exports now
+  conform to the aim stated above so you might need to add `hash_types::` to your paths for the more
+  esoteric hash types.
+
+## sighash
+
+We moved around and renamed a bunch of types to do with sighashes. In the `sighash` module, along
+with the `SighashCache` we now have the hopefully clearly named:
+
+- `LegacySighash`
+- `SegwitV0Sighash`
+- `TapSighash`
+- `EcdsaSighashType`
+- `TapSighashType`
+
+Signatures are now in their respective modules (`ecdsa` for legacy and segwit v0):
+
+- `taproot::Signature`
+- `ecdsa::Signature`
+
+## Lock types
+
+There are now two lock times, one for absolute locks (CLTV) and one for relative locks (CSV). We
+export the `absolute` and `relative` modules at the crate root so you can either import them from
+their or `use bitcoin::locktime::{absolute, relative};` if that's clearer. We expect locks to be
+used as `absolute::LockTime`.
+
+## Script changes
+
+We added a `ScriptBuf` alongside `Script`, these two now represent owned and borrowed types the same
+way that `PathBuf` and `Path` do (or `String` and `str`) i.e., we pass `&Script` if we only need to
+borrow.
+
+This change is quite invasive but should hopefully help to improve APIs here and in crates depending
+on us.
```

Comment:
The script changes are quite disruptive so I would place them higher up. Generally, I think we should order the changes from most disruptive (less intuitive changes to often-used-items) to least.

The description also doesn't accurately say what changed. It should be something like:

> To optimize efficiency of working with borrowed scripts we renamed `Script` to `ScriptBuf` and added an unsized `Script`. It works just like `PathBuf` and `Path` (and other such types) from `std`. The API tries to resemble those types as much as reasonable (deref coercions etc.), so it should be intuitive. Methods in the library that previously took `&Script` (which is now `ScriptBuf`) take the unsized `&Script` now.
>
> Additionally, we changed the type accepted by the `push_slice` method to be another unsized newtype: `PushBytes` and it's owned counterpart `PushBytesBuf`. These types maintain the invariant of storing at most 2^32-1 bytes - the maximum one can push into script. Previously the method would panic if you attempted to do it (and it wasn't documented, sorry about that). Now you can either handle it explicitly using `TryFrom` or just pass a known-length array (implemented for arrays up to 73 bytes).
>
> Types that are commonly pushed into script (serialized signatures, public keys...) implement `AsRef<PushBytes>` so you can pass those directly as well. You can also implement `AsRef` for your types so they can be pushed directly.

Reply:
Used vebatim.